Avoid redundant calls to signalKeyAsReady

yangbodong22011 opened the pull request:
Hi, This PR avoid invalid signalKeyAsReady.

[Edit] by @guybe7 advise
Avoid redundant calls to signalKeyAsReady
signalKeyAsReady has some overhead (namely dictFind) so we should
only call it when a client is blocked on the relevant type (BLOCKED_*)

[Edit]
The benefits of refactoring are as follows:
  1) The judgment like `val->type == OBJ_LIST` can be cancelled in dbAdd
  the code is clearer.
  2) By judging `server.blocked_clients_by_type[btype]`, avoid the cost
  of invalid calling `dictFind(db->blocking_keys)`.

madolson commented:
@yangbodong22011 this is a sensible change, but I think a slight refactoring might be in order here. I think the responsibility of checking if there are actually clients to unblock should be within signalKeyAsReady(). If we extend this function to include an optional type, we can do the check if we need to bother doing the update. This also fixes the two other cases where we don't do this check.

yangbodong22011 commented:
@madolson Good suggestion. I have finished the modification. Please review it again when you have time.

madolson commented on src/blocked.c:
```diff
@@ -634,16 +634,29 @@ void unblockClientWaitingData(client *c) {
     readyList *rl;
 
+    /* If this type is not blocked, just return */
+    if (!server.blocked_clients_by_type[btype]) return;
```
A minor suggestion. If this function were to take the object type instead of the blocked type and we do the conversion in this function, it would move more of the blocking code in blocked.c. You could omit the getBlockedTypeByType from server.h as well. It might add some extra overhead, but I think will help with maintainability.

yangbodong22011 replied:
@madolson Thanks, already updated.

oranagra commented:
@yangbodong22011 can you please update the commit comment (or more importantly the PR top comment which is used when this is squash-merged).

in bugfix commits please try to always describe the bug, in what scenario does it trigger and what are the consequences (what damage it does).

also, if the commit ended up doing some refactoring or other changes which where not part of the bugfix (e.g. cleanup for better readability or anything else), better mention that too (so it's easy to later tell what was done for each reason).

thanks.

yangbodong22011 commented:
@madolson @oranagra Thanks for pointing out my problem.
I modified the code and updated the PR commit message and conversation message. Regarding this PR commit message, if you have any suggestions, please suggest.

guybe7 commented:
for a commit comment, how about:
```
Avoid redundant calls to signalKeyAsReady

signalKeyAsReady has some overhead (namely dictFind) so we should
only call it when a client is blocked on the relevant type (BLOCKED_*)
```

oranagra commented on src/blocked.c:
```diff
@@ -634,16 +634,30 @@ void unblockClientWaitingData(client *c) {
+void signalKeyAsReady(redisDb *db, robj *key, int type) {
     readyList *rl;
 
+    /* If this type is not blocked, just return */
+    int btype = getBlockedTypeByType(type);
+    if (!server.blocked_clients_by_type[btype]) return;
```
```suggestion
    /* If no clients are blocked on this type, just return */
    int btype = getBlockedTypeByType(type);
    if (btype == BLOCKED_NONE || !server.blocked_clients_by_type[btype])
        return;
```

i know blocked_clients_by_type[0] is unused, but i think this way is better.
